### PR TITLE
added tests

### PR DESCRIPTION
**Pull request by Inve1951**
added a couple tests
i'm aware not all in there is ideal but please point me in the right direction

i thought to maybe move `hasActualAllExpectedPropsAndAreTheyEqual` to the top, ~~give it array support~~ (already has), and use it for all the tests

### CODE REVIEW

**Comment by GeoffreyBooth** [comment]
Could `hasActualAllExpectedPropsAndAreTheyEqual` be replaced by Node's `assert.deepStrictEqual`?

**Comment by Inve1951** [comment]
>Could hasActualAllExpectedPropsAndAreTheyEqual be replaced by Node’s assert.deepStrictEqual?

no, they work differently:
```coffee
actual = a: 1, b: 2
expected = a: 1   # don't care about `b`, ignore it

hasActualAllExpectedPropsAndAreTheyEqual actual, expected
# undefined

deepStrictEqual actual, expected   # throws
# AssertionError: { a: 1, b: 2 } deepStrictEqual { a: 1 }
```
`deepStrictEqual` even compares prototypes and symbols etc whereas `hasActualAllExpectedPropsAndAreTheyEqual` only checks enumerable keys of `expected`.

do those tests look alright so far? i got a feeling that i’m doing something awfully wrong.
also i couldn’t find some nodes, e.g. `Extends`. do they not exist?

**Comment by Inve1951** [comment]
@GeoffreyBooth I believe this is ready for a review.

File: test/abstract_syntax_tree.coffee [hunk 1]
```diff
@@ -10,207 +10,595 @@ getExpressions = (code) ->
...
+# Flag array for loose comparision. See above code/comment.
+looseArray = (arr) ->
+  Object.defineProperty arr, 'loose',
+    value: yes
+    enumerable: no
+  arr
+
+testExpressions = (code, expected) ->
+  ast = getExpressions code
+  return console.log ast unless expected?
```
Comment (GeoffreyBooth) [inline test/abstract_syntax_tree.coffee]: I think I added these `console.log` lines as helpers for writing the tests, so I could figure out what `expected` should be. Now that all the tests are written we don’t need them anymore.

File: test/abstract_syntax_tree.coffee [hunk 2]
```diff
@@ -10,207 +10,595 @@ getExpressions = (code) ->
...
+  deepStrictEqualExpectedProps ast, expected
+
+testExpression = (code, expected) ->
+  ast = getExpression code
+  return console.log ast unless expected?
```
Comment (GeoffreyBooth) [inline test/abstract_syntax_tree.coffee]: Here’s the other one.

File: test/abstract_syntax_tree.coffee [hunk 3]
```diff
@@ -10,207 +10,595 @@ getExpressions = (code) ->
...
+    ]
+
+  check 'Array length matches', doesNotThrow,
+    x: [1, 2, 3]
+
+  check 'Array prop mismatch', throws,
+    x: [3, 2, 1]
+
+  check 'Partial object comparison', doesNotThrow,
+    a: c: 2
```
Comment (GeoffreyBooth) [inline test/abstract_syntax_tree.coffee]: This line looks confusing. Is it:

```coffee
a:
  c: 2
```
?

File: test/abstract_syntax_tree.coffee [hunk 4]
```diff
@@ -10,207 +10,595 @@ getExpressions = (code) ->
...
 test "AST as expected for Value node", ->
-  expression = getExpression 'for i in [] then i'
-  value = expression.body
-  eq value.type, 'Value'
-  # eq value.value, 'i'
+  testExpression 'for i in [] then i',
+    body:
+      type: 'Value'
+      isDefaultValue: no
+      base: value: 'i'
```
Comment (GeoffreyBooth) [inline test/abstract_syntax_tree.coffee]: In general can we avoid implicit objects as object values? It seems like a recipe for trouble later.
```coffee
base:
  value: 'i'
```

File: test/abstract_syntax_tree.coffee [hunk 5]
```diff
@@ -10,207 +10,595 @@ getExpressions = (code) ->
...
+        type: 'Super'
+        accessor: type: 'Access'
+    ]
 
 test "AST as expected for RegexWithInterpolations node", ->
-  # TODO
+  testExpression '///^#{flavor}script$///',
+    type: 'RegexWithInterpolations'
+
+  # TODO: Shouldn't there be more info we can check for?
```
Comment (GeoffreyBooth) [inline test/abstract_syntax_tree.coffee:332]: Could you check for the body of the regex?

File: test/abstract_syntax_tree.coffee [hunk 6]
```diff
@@ -10,207 +10,595 @@ getExpressions = (code) ->
...
+      value: type: 'Code'
 
 test "AST as expected for ExportDefaultDeclaration node", ->
-  # TODO
+  testExpression 'export default class',
+    type: 'ExportDefaultDeclaration'
+    clause: type: 'Class'
 
 test "AST as expected for ExportAllDeclaration node", ->
   # TODO
```
Comment (GeoffreyBooth) [inline test/abstract_syntax_tree.coffee]: We could add this one, couldn’t we?

File: test/abstract_syntax_tree.coffee [hunk 7]
```diff
@@ -219,40 +607,101 @@ test "AST as expected for ModuleSpecifierList node", ->
...
```
Comment (GeoffreyBooth) [inline test/abstract_syntax_tree.coffee]: I think `ModuleSpecifierList` is a parent class that would never be output on its own. If that’s correct, we shouldn’t need to test it (well, we couldn’t, actually).

File: test/abstract_syntax_tree.coffee [hunk 8]
```diff
@@ -219,40 +607,101 @@ test "AST as expected for ModuleSpecifierList node", ->
...
+    params: [
+      type: 'Param'
+      splat: yes
+      name: value: 'a'
+    ]
+
+  # TODO: Test object splats.
 
 test "AST as expected for Expansion node", ->
   # TODO
```
Comment (GeoffreyBooth) [inline test/abstract_syntax_tree.coffee]: Can we add this one, and the following?

File: test/abstract_syntax_tree.coffee [hunk 9]
```diff
@@ -261,34 +710,246 @@ test "AST as expected for Elision node", ->
...
+      expressions: [
+        originalValue: ''
+      ,
+        type: 'Interpolation'
+        expression:
+          type: 'Value'
+          base: value: 'o'
+      ,
+        originalValue: '/'
+      ]
```
Comment (GeoffreyBooth) [inline test/abstract_syntax_tree.coffee]: For strings, we will want to test (eventually) that the AST outputs the string delimiter (single quote or double quote). I assume that’s not output yet, so maybe add a TODO?

File: test/abstract_syntax_tree.coffee [hunk 10]
```diff
@@ -261,34 +710,246 @@ test "AST as expected for Elision node", ->
...
+      {type: 'IdentifierLiteral', value: 'b'}
+      {type: 'IdentifierLiteral', value: 'c'}
+      {type: 'Value', base: value: 'c'}
+    ]
+    otherwise:
+      type: 'Value'
+      base: value: '42'
+      isDefaultValue: no
+
+  # TODO: File issue for compile error when using `then` or `;` where `\n` is rn.
```
Comment (GeoffreyBooth) [inline test/abstract_syntax_tree.coffee]: ?

**Review by GeoffreyBooth — CHANGES_REQUESTED** [review]
Great work so far! All minor notes.

**Comment by GeoffreyBooth** [comment]
> do those tests look alright so far? i got a feeling that i’m doing something awfully wrong.
also i couldn’t find some nodes, e.g. `Extends`. do they not exist?

I think you’re on the right track. In general we probably want to check more values of things, like the value of a string or the name of a function/`Call`. We especially want to check the unusual attributes of nodes, like `soak` etc., to make sure that they exist and are what we’d expect.

As for `Extends`, it may be that it’s just never output in the final AST because it becomes included in some larger node—in that case, probably a `Class`. If that’s the case, we should try to find it inside that larger node and validate it there.

File: test/abstract_syntax_tree.coffee (hunk 1 above)
Comment (Inve1951) [reply test/abstract_syntax_tree.coffee]: I added those logs and would leave them there for future edits. They are very convenient.

File: test/abstract_syntax_tree.coffee (hunk 3 above)
Comment (Inve1951) [reply test/abstract_syntax_tree.coffee]: Yes it is. To me those aren't confusing at all. It's just coffeescript.

File: test/abstract_syntax_tree.coffee (hunk 4 above)
Comment (Inve1951) [reply test/abstract_syntax_tree.coffee]: Well, we could. But not only would it make the file much longer, it would also make it less readable IMO.

File: test/abstract_syntax_tree.coffee (hunk 5 above)
Comment (Inve1951) [reply test/abstract_syntax_tree.coffee:332]: That's what i wanted to do but it seems the information is missing. Here's the full AST from `getExpressions`:
```js
[
  {
    type: 'RegexWithInterpolations',
    loc: {
      start: {line: 0, column: 0},
      end: {line: 0, column: 22}
    }
  }
]
```

File: test/abstract_syntax_tree.coffee (hunk 6 above)
Comment (Inve1951) [reply test/abstract_syntax_tree.coffee]: Unfortunately I have no idea what it could be.

File: test/abstract_syntax_tree.coffee (hunk 7 above)
Comment (Inve1951) [reply test/abstract_syntax_tree.coffee]: I was under the same impression. It's likely the base class for `ExportSpecifierList` and `ImportSpecifierList`. Same would go for the other `Module` nodes.

File: test/abstract_syntax_tree.coffee (hunk 8 above)
Comment (Inve1951) [reply test/abstract_syntax_tree.coffee]: No idea what they could be.

File: test/abstract_syntax_tree.coffee (hunk 9 above)
Comment (Inve1951) [reply test/abstract_syntax_tree.coffee]: That information is available in the `quote` property and is even checked here. I added the check to `StringLiteral` now.

File: test/abstract_syntax_tree.coffee (hunk 10 above)
Comment (Inve1951) [reply test/abstract_syntax_tree.coffee]: From playing around for a bit it seemed impossible to enter the `Switch` body without an actual linebreak. From my understanding a semicolon or the `then` keyword should allow for it.

**Comment by Inve1951** [comment]
Replied to your comments.

>I think you’re on the right track. In general we probably want to check more values of things, like the value of a string or the name of a function/Call. We especially want to check the unusual attributes of nodes, like soak etc., to make sure that they exist and are what we’d expect.

Absolutely. But I feel that the current AST generator isn't quite there yet. Some information seems missing (`RegexWithInterpolations`) or does not keep track of the originally used code (`unless`). And some I can't make sense of (`return` prop on `For` nodes).

>As for Extends, it may be that it’s just never output in the final AST because it becomes included in some larger node—in that case, probably a Class. If that’s the case, we should try to find it inside that larger node and validate it there.

It seems that `Extends` does not have it's own node. However extending a class leaves the `Class` node with a `parent` property. That's what the test currently checks for.

<br>
If i followed correctly then someone (@helixbass?) is still working on AST output. The tests I wrote in this PR so far all pass on the current compiler output, which i expect will still change as work on the AST generator progresses. I'd rather not make the tests too strict just yet before the AST is final.

**Comment by helixbass** [comment]
@Inve1951 yup I have been working on AST generation on a separate [branch](https://github.com/helixbass/copheescript/tree/prettier)

Wherever possible, I've been making the structure of our AST nodes match the corresponding [Babel AST nodes](https://github.com/babel/babel/blob/master/packages/babel-parser/ast/spec.md), as there are lots of upsides to compatibility. Then when there are Coffeescript-specific things, I've exposed additional fields on some node types and created a couple new node types that don't have a comparable node type in JS

@GeoffreyBooth and I discussed a strategy for getting the work I've been doing merged into our `ast` branch where I can open individual PRs for each node type's AST support and make sure that the corresponding AST tests reflect the expected shape of the AST nodes (including any edge cases)

So basically I'll be prepared to pick up the existing tests on this branch in whatever state they're in and tweak them/add to them as I go

If  you're curious, I'd recommend [astexplorer.net](https://astexplorer.net) as a great way of exploring the different AST node types - choosing the `babylon7` parser should allow you to inspect the Babel-style AST nodes for any ES6+ code. And/or you can check out my branch and try dumping ASTs for Coffeescript source using the `--ast` command-line flag (let me know if you need help getting this working)

**Comment by Inve1951** [comment]
@helixbass i cloned your prettier branch to see how much the AST differs. However it fails to build the compiler as it can't handle the object splats. I've tried installing `babel-plugin-transform-object-rest-spread`, adding it to `.babelrc` and even to `coffeescript.coffee`'s `compile` function but it still fails to `cake build:full` when encountering a splat. No idea how to resolve this (besides upgrading node).

**Comment by helixbass** [comment]
@Inve1951 ah yeah I recall the discussion that there shouldn't be object spread syntax in the Coffeescript codebase so as to still support Node 6.x. So I'll have to go through and update places where I'm using it

In the meantime, if upgrading Node isn't an option, I'm not sure exactly how you'd have to configure Babel to get it to post-process the compiled `lib/coffeescript/*.js` files. If you want to get really dirty, you could probably hack it into the build process manually eg by adding a Babel invocation at the end of `buildExceptParser()` in `Cakefile`

**Comment by GeoffreyBooth** [comment]
Please merge `master` into your branches, I added a fix for dependencies.

@helixbass I’ve been using `Object.assign` instead of object splats, if you look around the codebase you can find some examples.

Building the compiler doesn’t use Babel, other than for building the browser compiler. On `master` it builds from Node 4 and up. Object splats are supported in Node 8 and up.

File: test/abstract_syntax_tree.coffee (hunk 4 above)
Comment (GeoffreyBooth) [reply test/abstract_syntax_tree.coffee]: Unless I’m mistaken, it’s not the coding style of the rest of the codebase. We try to keep things consistent.

File: test/abstract_syntax_tree.coffee (hunk 5 above)
Comment (GeoffreyBooth) [reply test/abstract_syntax_tree.coffee:332]: Yeah, I think this is the kind of thing that @helixbass is adding in his work.

File: test/abstract_syntax_tree.coffee (hunk 6 above)
Comment (GeoffreyBooth) [reply test/abstract_syntax_tree.coffee]:
```coffee
export * from "module-name"
```

File: test/abstract_syntax_tree.coffee (hunk 7 above)
Comment (GeoffreyBooth) [reply test/abstract_syntax_tree.coffee]: If it’s never output in the AST, we don’t need a TODO for it.

File: test/abstract_syntax_tree.coffee (hunk 9 above)
Comment (GeoffreyBooth) [reply test/abstract_syntax_tree.coffee]: Ah, sorry I didn’t notice. Good catch.

File: test/abstract_syntax_tree.coffee (hunk 8 above)
Comment (GeoffreyBooth) [reply test/abstract_syntax_tree.coffee]: `Expansion` is defined in a comment as “Used to skip values inside an array destructuring (pattern matching) or parameter list.” I wonder if this node is even used anymore, now that we use ES destructuring.

Comment (GeoffreyBooth) [reply test/abstract_syntax_tree.coffee]: `Elision` is just `[,a, , , b, , c, ,]` (the skipped elements). Is that not output in the AST?

**Comment by Inve1951** [comment]
Updated per your comments. These node types are still left without any tests: `ImportNamespaceSpecifier`, `Expansion`

**Review by GeoffreyBooth — APPROVED** [review]
